File: web/src/engine/formula.ts
// Safe arithmetic evaluator — a tokenizer + recursive-descent parser. NO eval,
// NO Function constructor, no property access. Mirrors engine/ccscores/formula.py
// so both engines agree on every golden vector.

export class EngineError extends Error {}

export function roundHalfUp(x: number): number {
  // floor(x + 0.5) === JS Math.round for finite x; matches the Python engine.
  return Math.floor(x + 0.5);
}

const FUNCS: Record<string, { fn: (...a: number[]) => number; arity: number | null }> = {
  ln: { fn: Math.log, arity: 1 },
  exp: { fn: Math.exp, arity: 1 },
  log10: { fn: Math.log10, arity: 1 },
  sqrt: { fn: Math.sqrt, arity: 1 },
  floor: { fn: Math.floor, arity: 1 },
  ceil: { fn: Math.ceil, arity: 1 },
  abs: { fn: Math.abs, arity: 1 },
  round: { fn: roundHalfUp, arity: 1 },
  min: { fn: Math.min, arity: null },
  max: { fn: Math.max, arity: null },
  clamp: { fn: (x, lo, hi) => Math.max(lo, Math.min(hi, x)), arity: 3 },
};

type Tok =
  | { t: "num"; v: number }
  | { t: "id"; v: string }
  | { t: "op"; v: string }
  | { t: "lparen" }
  | { t: "rparen" }
  | { t: "comma" };

function tokenize(src: string): Tok[] {
  const toks: Tok[] = [];
  let i = 0;
  while (i < src.length) {
    const c = src[i];
    if (c === " " || c === "\t" || c === "\n") {
      i++;
      continue;
    }
    if (c >= "0" && c <= "9") {
      let j = i;
      while (j < src.length && /[0-9.eE+\-]/.test(src[j])) {
        // allow exponent sign only right after e/E
        if ((src[j] === "+" || src[j] === "-") && !(src[j - 1] === "e" || src[j - 1] === "E")) break;
        j++;
      }
      const num = Number(src.slice(i, j));
      if (Number.isNaN(num)) throw new EngineError(`invalid number near '${src.slice(i, j)}'`);
      toks.push({ t: "num", v: num });
      i = j;
      continue;
    }
    if (/[a-zA-Z_]/.test(c)) {
      let j = i;
      while (j < src.length && /[a-zA-Z0-9_]/.test(src[j])) j++;
      toks.push({ t: "id", v: src.slice(i, j) });
      i = j;
      continue;
    }
    if ("+-*/%^".includes(c)) {
      toks.push({ t: "op", v: c });
      i++;
      continue;
    }
    if (c === "(") {
      toks.push({ t: "lparen" });
      i++;
      continue;
    }
    if (c === ")") {
      toks.push({ t: "rparen" });
      i++;
      continue;
    }
    if (c === ",") {
      toks.push({ t: "comma" });
      i++;
      continue;
    }
    throw new EngineError(`unexpected character '${c}' in formula`);
  }
  return toks;
}

export function evaluate(expr: string, variables: Record<string, unknown>): number {
  const toks = tokenize(expr);
  let pos = 0;

  const peek = () => toks[pos];
  const next = () => toks[pos++];
  const isOp = (...ops: string[]) => {
    const tk = toks[pos];
    return tk?.t === "op" && ops.includes(tk.v);
  };

  function coerce(name: string): number {
    const v = variables[name];
    if (typeof v === "boolean") return v ? 1 : 0;
    if (typeof v === "number") return v;
    if (v === undefined) throw new EngineError(`formula references unknown variable '${name}'`);
    throw new EngineError(`formula variable '${name}' is not numeric`);
  }

  function parseExpr(): number {
    let left = parseTerm();
    while (isOp("+", "-")) {
      const op = (next() as { t: "op"; v: string }).v;
      const right = parseTerm();
      left = op === "+" ? left + right : left - right;
    }
    return left;
  }

  function parseTerm(): number {
    let left = parsePower();
    while (isOp("*", "/", "%")) {
      const op = (next() as { t: "op"; v: string }).v;
      const right = parsePower();
      left = op === "*" ? left * right : op === "/" ? left / right : left % right;
    }
    return left;
  }

  function parsePower(): number {
    const base = parseUnary();
    if (isOp("^")) {
      next();
      return Math.pow(base, parsePower()); // right-associative
    }
    return base;
  }

  function parseUnary(): number {
    if (isOp("+", "-")) {
      const op = (next() as { t: "op"; v: string }).v;
      const val = parseUnary();
      return op === "-" ? -val : val;
    }
    return parsePrimary();
  }

  function parsePrimary(): number {
    const tk = peek();
    if (!tk) throw new EngineError("unexpected end of formula");
    if (tk.t === "num") {
      next();
      return tk.v;
    }
    if (tk.t === "lparen") {
      next();
      const v = parseExpr();
      if (next()?.t !== "rparen") throw new EngineError("missing ')' in formula");
      return v;
    }
    if (tk.t === "id") {
      next();
      if (peek()?.t === "lparen") {
        const spec = FUNCS[tk.v];
        if (!spec) throw new EngineError(`call to disallowed function '${tk.v}'`);
        next(); // (
        const args: number[] = [];
        if (peek()?.t !== "rparen") {
          args.push(parseExpr());
          while (peek()?.t === "comma") {
            next();
            args.push(parseExpr());
          }
        }
        if (next()?.t !== "rparen") throw new EngineError("missing ')' in function call");
        if (spec.arity !== null && args.length !== spec.arity)
          throw new EngineError(`function '${tk.v}' expects ${spec.arity} args, got ${args.length}`);
        return spec.fn(...args);
      }
      return coerce(tk.v);
    }
    throw new EngineError(`unexpected token in formula`);
  }

  const result = parseExpr();
  if (pos !== toks.length) throw new EngineError("trailing tokens in formula");
  return result;
}

File: engine/ccscores/formula.py
"""Safe arithmetic expression evaluator for score formulas.

Parses an expression with Python's `ast` module and walks a strict whitelist of
node types — NO `eval`, no attribute access, no arbitrary calls. Names resolve
to input values; booleans coerce to 1/0. The grammar deliberately mirrors what
the TypeScript engine implements so both stay in lock-step.

Allowed functions: ln, exp, log10, sqrt, min, max, round, floor, ceil, abs,
clamp(x, lo, hi).
"""

from __future__ import annotations

import ast
import math
from typing import Mapping

from .errors import EngineError


def _clamp(x: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, x))


def round_half_up(x: float) -> float:
    """Round half away-from-zero via floor(x+0.5), matching JS Math.round.

    Used by both engines so cross-implementation golden vectors agree exactly
    (Python's built-in round() is banker's rounding and would diverge)."""
    return float(math.floor(x + 0.5))


# name -> (callable, arity or None for variadic min/max)
_FUNCS = {
    "ln": (math.log, 1),
    "exp": (math.exp, 1),
    "log10": (math.log10, 1),
    "sqrt": (math.sqrt, 1),
    "floor": (lambda x: float(math.floor(x)), 1),
    "ceil": (lambda x: float(math.ceil(x)), 1),
    "abs": (abs, 1),
    "round": (round_half_up, 1),
    "min": (min, None),
    "max": (max, None),
    "clamp": (_clamp, 3),
}

_BINOPS = {
    ast.Add: lambda a, b: a + b,
    ast.Sub: lambda a, b: a - b,
    ast.Mult: lambda a, b: a * b,
    ast.Div: lambda a, b: a / b,
    ast.Pow: lambda a, b: a ** b,
    ast.Mod: lambda a, b: a % b,
}


def _coerce(value: object, name: str) -> float:
    if isinstance(value, bool):
        return 1.0 if value else 0.0
    if isinstance(value, (int, float)):
        return float(value)
    raise EngineError(f"formula variable '{name}' is not numeric: {value!r}")


def evaluate(expr: str, variables: Mapping[str, object]) -> float:
    """Evaluate `expr` over `variables`. Raises EngineError on anything unsafe."""
    try:
        tree = ast.parse(expr, mode="eval")
    except SyntaxError as exc:  # pragma: no cover - guarded by schema in practice
        raise EngineError(f"invalid formula syntax: {expr!r} ({exc})") from exc

    def walk(node: ast.AST) -> float:
        if isinstance(node, ast.Expression):
            return walk(node.body)
        if isinstance(node, ast.Constant):
            if isinstance(node.value, bool):
                return 1.0 if node.value else 0.0
            if isinstance(node.value, (int, float)):
                return float(node.value)
            raise EngineError(f"unsupported literal in formula: {node.value!r}")
        if isinstance(node, ast.Name):
            if node.id not in variables:
                raise EngineError(f"formula references unknown variable '{node.id}'")
            return _coerce(variables[node.id], node.id)
        if isinstance(node, ast.UnaryOp):
            if isinstance(node.op, ast.USub):
                return -walk(node.operand)
            if isinstance(node.op, ast.UAdd):
                return +walk(node.operand)
            raise EngineError("unsupported unary operator in formula")
        if isinstance(node, ast.BinOp):
            op = _BINOPS.get(type(node.op))
            if op is None:
                raise EngineError(f"unsupported operator in formula: {type(node.op).__name__}")
            return op(walk(node.left), walk(node.right))
        if isinstance(node, ast.Call):
            if not isinstance(node.func, ast.Name) or node.keywords or node.func.id not in _FUNCS:
                fn = getattr(node.func, "id", "?")
                raise EngineError(f"call to disallowed function '{fn}' in formula")
            func, arity = _FUNCS[node.func.id]
            args = [walk(a) for a in node.args]
            if arity is not None and len(args) != arity:
                raise EngineError(f"function '{node.func.id}' expects {arity} args, got {len(args)}")
            return float(func(*args))
        raise EngineError(f"disallowed expression element: {type(node).__name__}")

    return walk(tree)
